src/Microsoft.OData.Client/DataServiceExtensions.cs

```diff
@@ -0,0 +1,156 @@
+        {
+            MethodCallExpression countMethodExpr;
+
+            if (source.Provider.GetType().Equals(typeof(DataServiceQueryProvider)))
+            {
+                var currentMethod = (MethodInfo)MethodBase.GetCurrentMethod();
+                MethodInfo methodInfo = currentMethod.MakeGenericMethod(typeof(TSource), typeof(TTarget));
+
+                countMethodExpr = Expression.Call(null,
+                    methodInfo,
+                    new Expression[] { source.Expression, selector });
+            }
+            else
+            {
+                // Provide a default implementation...
+                // In the event that a developer who adds a reference to the library invokes CountDistinct() as follows:
+                // - new List<int> { 1, 2, 1 }.AsQueryable().CountDistinct(d => d)
+
+                // Method: Select<TSource,TResult>(IQueryable<TSource>, Expression<Func<TSource,TResult>>)
+                MethodInfo selectMethod = GetSelectMethod();
+                // Method: Distinct<TSource>(IQueryable<TSource>)
+                MethodInfo distinctMethod = GetDistinctMethod(typeof(Queryable), typeof(IQueryable<>));
+                // Method: Count<TSource>(IQueryable<TSource>)
+                MethodInfo countMethod = GetCountMethod(typeof(Queryable), typeof(IQueryable<>));
+
+                // Select(d => d.Prop)
+                MethodCallExpression selectMethodExpr = Expression.Call(null,
+                    selectMethod.MakeGenericMethod(new Type[] { source.ElementType, selector.Body.Type }),
+                    new[] { source.Expression, Expression.Quote(selector) });
+
+                // Select(d => d.Prop).Distinct()
+                MethodCallExpression distinctMethodExpr = Expression.Call(null,
+                    distinctMethod.MakeGenericMethod(new Type[] { selector.Body.Type }),
+                    new[] { selectMethodExpr });
+
+                // Select(d => d.Prop).Distinct().Count()
+                countMethodExpr = Expression.Call(null,
+                    countMethod.MakeGenericMethod(new Type[] { selector.Body.Type }),
+                    new[] { distinctMethodExpr });
+            }
+
+            return source.Provider.Execute<int>(countMethodExpr);
+        }
+
+        /// <summary>
+        /// Returns the distinct count of elements in a sequence after applying the projection function to each element.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source"/></typeparam>
+        /// <typeparam name="TTarget">The type returned by the projection function represented in <paramref name="selector"/>.</typeparam>
+        /// <param name="source">A sequence of values of type <typeparamref name="TSource"/>.</param>
+        /// <param name="selector">A projection function to apply to each element.</param>

```

Comment:
For the `CountDistinct` methods, the `selector` params description should be different. One is an `Expression`, the other is a `Func`.

Reply:
@KenitoInc Note that in the official docs for `Queryable` and `Enumerable` methods, the descriptions don't differ much

`selector
Func<TSource,TResult>` -  [A transform function to apply to each element.](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.select?view=net-5.0#System_Linq_Enumerable_Select__2_System_Collections_Generic_IEnumerable___0__System_Func___0___1__)
`selector
Expression<Func<TSource,TResult>>` - [A projection function to apply to each element.](https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.select?view=net-5.0#System_Linq_Queryable_Select__2_System_Linq_IQueryable___0__System_Linq_Expressions_Expression_System_Func___0___1___)